src/org/traccar/protocol/NavisProtocolDecoder.java

```diff
@@ -272,70 +316,426 @@ private Object processArray(DeviceSession deviceSession, Channel channel, ByteBu
+                    }
+                    break;
+                case 36:
+                    position.set(Position.KEY_HOURS, buf.readUnsignedIntLE() * 1000);
+                    break;
+                case 44:
+                case 45:
+                case 46:
+                case 47:
+                case 48:
+                case 49:
+                case 50:
+                case 51:
+                    position.set(Position.PREFIX_TEMP + (i - 43), buf.readByte());
+                    break;
+                case 68:  // CAN Speed
+                    position.set("can-speed", buf.readUnsignedByte());
+                    break;
+                // FLEX 2.0
+                case 69:
+                    int satVisible = 0;
+                    for (int k = 0; k < 8; k++) {
+                        switch (k) {
+                            case 0:
+                                int satVisibleGLONASS = buf.readUnsignedByte();
+                                position.set("sat-visible-glonass", satVisibleGLONASS);
+                                satVisible += satVisibleGLONASS;
+                                break;
+                            case 1:
+                                int satVisibleGPS = buf.readUnsignedByte();
+                                position.set("sat-visible-gps", satVisibleGPS);
+                                satVisible += satVisibleGPS;
+                                break;
+                            default:
+                                // Don't detalize Galileo, Compass, Beidou, DORIS, IRNSS, QZSS
+                                satVisible += buf.readUnsignedByte();
+                                break;
+                }
+                    }
+                    position.set(Position.KEY_SATELLITES_VISIBLE, satVisible);
+                    break;
+                case 70:
+                    position.set(Position.KEY_HDOP, buf.readUnsignedByte() * 0.1);
+                    position.set(Position.KEY_PDOP, buf.readUnsignedByte() * 0.1);
+                    break;
+                default:
+                    skipFlexField(i, buf);
+                    break;
+            }
+        }
+
+        return new ParseResult(index, position);
+    }
+
+    private ParseResult parseFlex20Position(DeviceSession deviceSession, ByteBuf buf) {
+        Position position = new Position(getProtocolName());
+
+        position.setDeviceId(deviceSession.getDeviceId());
+
+        long index = 0;
+        int length = buf.readUnsignedShort();
+
+        // Check buffer size and struct version
+        if (length <= buf.readableBytes() && buf.readUnsignedByte() == 0x0A) {
+            buf.skipBytes(1);  // Length of static part
+
+            index = buf.readUnsignedIntLE();
+            position.set(Position.KEY_INDEX, index);
+
+            position.set(Position.KEY_EVENT, buf.readUnsignedShortLE());
+            buf.skipBytes(4); // Event time
```

Comment:
For fields skipping we usually use `readXxx` or `readUnsignedXxx`. Please avoid `skipBytes` unless it's an array or something.

Reply:
Fixed